### PR TITLE
Core/Spells: SPELL_AURA_48 implementation

### PR DESCRIPTION
**Pull request by Faq**
by xvwyh

**Changes proposed:**
SPELL_AURA_48 implementation

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #  (insert issue tracker number)
#23204

**Tests performed:** (Does it build, tested in-game, etc.)

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] 
- [ ]

### CODE REVIEW

**Comment by Killyana** [comment]
I tested the PR, but the spell still don't deal any damage.

**Comment by xvwyh** [comment]
Case and point for why I'm not making PRs. These were enough for my ancient fork of TC, I cannot know if more changes are needed for TC. @Killyana, I sincerely hope you tested this PR and not the patch you commented with in #23204, because you omitted a crucial detail, and you never responded to my correction.

Can't do some basic debugging to pinpoint exactly which spell is failing, instead of simply stating that it deals no damage? Is 67200/68829 being cast by server? Does the cast succeed or does `CheckCast` fail? With what result? Is 67199/67450 being cast by server? Does the cast succeed? Does it fail because of the range check (67200/68829 can launch missiles at 50000 yds distance, but the 67199/67450 that they're triggering on hit only has 150 yds range)? Something else?

**Comment by Killyana** [comment]
I tested this PR and what ever the distance it doesn't deal damage also other player don't see thee missiles.

**Comment by xvwyh** [comment]
Oh jolly, this PR missed another handler. Sigh, I understand that everyone wants complete patches, but reading shouldn't be that hard...

> WorldSession::HandleCastSpellOpcode (if the aura is on player)
> WorldSession::HandlePetCastSpellOpcode (if the aura is on vehicle)

**Comment by xvwyh** [comment]
Here...
WorldSession::HandlePetCastSpellOpcode
```diff
         return;
     }
 
-    // do not cast not learned spells
-    if (!caster->HasSpell(spellId) || spellInfo->IsPassive())
-        return;
 
     SpellCastTargets targets;
     targets.Read(recvPacket, caster);
     HandleClientCastFlags(recvPacket, castFlags, targets);
 
+    TriggerCastFlags triggerCastFlags = TRIGGERED_NONE;
+
+    if (spellInfo->IsPassive())
+        return;
+
+    // do not cast not learned spells
+    if (!caster->HasSpell(spellId))
+    {
+        bool allow = false;
+
+        // allow casting of spells triggered by clientside periodic trigger auras
+        if (caster->HasAuraTypeWithTriggerSpell(SPELL_AURA_PERIODIC_TRIGGER_SPELL_FROM_CLIENT, spellId))
+        {
+            allow = true;
+            triggerCastFlags = TRIGGERED_FULL_MASK;
+        }
+
+        if (!allow)
+            return;
+    }
 
-    Spell* spell = new Spell(caster, spellInfo, TRIGGERED_NONE);
+    Spell* spell = new Spell(caster, spellInfo, triggerCastFlags);
     spell->m_cast_count = castCount; // probably pending spell cast
     spell->m_targets = targets;
```

**Comment by Killyana** [comment]
Yes, looks good.

**Comment by jackpoz** [comment]
@Faq could you update the PR with latest patch ?

**Comment by Faq** [comment]
Yes, finally victory over git.

**Comment by Killyana** [comment]
We still need: WorldSession::HandleCastSpellOpcode (if the aura is on player)

**Comment by xvwyh** [comment]
@Killyana, https://github.com/TrinityCore/TrinityCore/pull/24054/files#diff-059dfe52f1f1035b531419dab5ace957L388

**Comment by Killyana** [comment]
@xvwyh can you review the PR before we merge it?

**Comment by xvwyh** [comment]
If it works in-game, then LGTM code-wise.

File: src/server/game/Entities/Unit/Unit.cpp [hunk 1]
```diff
@@ -4503,6 +4503,14 @@ bool Unit::HasAuraTypeWithValue(AuraType auraType, int32 value) const
     return false;
 }
 
+bool Unit::HasAuraTypeWithTriggerSpell(AuraType auratype, uint32 triggerSpell) const
+{
+    for (auto&& aura : GetAuraEffectsByType(auratype))
```
Comment (ccrs) [inline src/server/game/Entities/Unit/Unit.cpp]: no auto for value iteration

Comment (Keader) [reply src/server/game/Entities/Unit/Unit.cpp]: wait, if i remember right, auto is allowed for iterators

Comment (Shauren) [reply src/server/game/Entities/Unit/Unit.cpp]: `auto` is allowed for iterators when you actually have the iterator object (so when you use `.begin()` and `.end()`, not in range style loop)

Comment (Faq) [reply src/server/game/Entities/Unit/Unit.cpp]: Like this?

`for (AuraEffect const* aura : GetAuraEffectsByType(auratype))`

Comment (Keader) [reply src/server/game/Entities/Unit/Unit.cpp]: Yes

Comment (Faq) [reply src/server/game/Entities/Unit/Unit.cpp]: Done

**Comment by jackpoz** [comment]
is this ready to be merged ?

**Comment by ccrs** [comment]
if the design is correct, the implementation looks good to me :+1:

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Pull request merged**